File: 34 Find First and Last Position of Element in Sorted Array/34.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] res = {-1, -1};
        if (nums.length == 0) return res;
        res[0] = get_index(nums, target, true);
        res[1] = get_index(nums, target, false);
        return res;
    }

    public int get_index(int[] nums, int target, boolean isLeft) {
        int start = 0;
        int end = nums.length - 1;
        while (start <= end) {
            int mid = start + (end - start) / 2;
            if (nums[mid] < target){
                start = mid + 1;
            } else if (nums[mid] > target){
                end = mid - 1;
            } else {
                if (isLeft) {
                    if (mid == 0 || nums[mid - 1] < target) {
                        return mid;
                    } else {
                        end = mid - 1;
                    }
                } else {
                    if (mid == nums.length - 1 || nums[mid + 1] > target) {
                        return mid;
                    } else {
                        start = mid + 1;
                    }
                }
            }
        }
        return -1;
    }
}

File: 34 Find First and Last Position of Element in Sorted Array/34.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
    let length = nums.length
    if (!length) return [-1, -1];
    return [get_index(nums, target, 0, length-1, true), get_index(nums, target, 0, length-1, false)]
};

var get_index = function(nums, target, start, end, isLeft) {
    if (start > end) {
        return -1;
    }
    let mid = start + ((end - start) >> 1)
    if (isLeft) {
        if (nums[mid] == target && (mid == 0 || nums[mid-1] < target)) {
            return mid;
        } else if (nums[mid] < target) {
            return get_index(nums, target, mid + 1, end, isLeft);
        } else {
            return get_index(nums, target, start, mid - 1, isLeft);
        }
    } else {
        if (nums[mid] == target && (mid == end || nums[mid+1] > target)) {
            return mid;
        } else if (nums[mid] > target) {
            return get_index(nums, target, start, mid - 1, isLeft);
        } else {
            return get_index(nums, target, mid + 1, end, isLeft);
        } 
    }
}